Sources/PackageCollections/Storage/SQLitePackageCollectionsStorage.swift

```diff
@@ -223,54 +322,61 @@ final class SQLitePackageCollectionsStorage: PackageCollectionsStorage, Closable
         }
     }
 
-    // TODO: this is PoC for search, need a more performant version of this
     func searchPackages(identifiers: [Model.CollectionIdentifier]? = nil,
                         query: String,
                         callback: @escaping (Result<Model.PackageSearchResult, Error>) -> Void) {
-        let queryString = query.lowercased()
-
         self.list(identifiers: identifiers) { result in
             switch result {
             case .failure(let error):
                 callback(.failure(error))
             case .success(let collections):
-                let collectionsPackages = collections.reduce([Model.CollectionIdentifier: [Model.Package]]()) { partial, collection in
-                    var map = partial
-                    map[collection.identifier] = collection.packages.filter { package in
-                        if package.repository.url.lowercased().contains(queryString) { return true }
-                        if let summary = package.summary, summary.lowercased().contains(queryString) { return true }
-                        if let keywords = package.keywords, (keywords.map { $0.lowercased() }).contains(queryString) { return true }
-                        return package.versions.contains(where: { version in
-                            if version.packageName.lowercased().contains(queryString) { return true }
-                            if version.products.contains(where: { $0.name.lowercased().contains(queryString) }) { return true }
-                            return version.targets.contains(where: { $0.name.lowercased().contains(queryString) })
-                        })
+                var matches = [(collection: Model.CollectionIdentifier, package: PackageIdentity)]()
+                do {
+                    let packageQuery = "SELECT collection_id_blob_base64, repository_url FROM \(Self.packagesFTSName) WHERE \(Self.packagesFTSName) MATCH ?;"
+                    try self.executeStatement(packageQuery) { statement in
+                        try statement.bind([.string(query)])
+                        
+                        while let row = try statement.step() {
+                            if let collectionData = Data(base64Encoded: row.string(at: 0)),
+                               let collection = try? self.decoder.decode(Model.CollectionIdentifier.self, from: collectionData) {
+                                matches.append((collection: collection, package: PackageIdentity(url: row.string(at: 1))))
+                            }
+                        }
                     }
-                    return map
+                } catch {
+                    return callback(.failure(error))
                 }
 
-                // compose result :p
+                let collectionDict = collections.reduce(into: [Model.CollectionIdentifier: Model.Collection]()) { result, collection in
+                    result[collection.identifier] = collection
+                }
```

Comment:
does this work? usually iirc you need to put the partial result into a var and then return it

Reply:
ah, using `into:` 👍